pkg/monitoring/vms/prometheus/prometheus.go

```diff
@@ -504,18 +514,49 @@ func updateVMIsPhase(nodeName string, vmis []*k6tv1.VirtualMachineInstance, ch c
 			continue
 		}
 		ch <- mv
+
 	}
 }
 
-func updateVMIEvictionBlocker(nodeName string, vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+func updateVMIMetrics(vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
 	for _, vmi := range vmis {
+		updateVMIEvictionBlocker(vmi, ch)
+		updateVMIPhaseTransitionTime(vmi, ch)
+	}
+
+}
+
+func updateVMIEvictionBlocker(vmi *k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+	mv, err := prometheus.NewConstMetric(
+		vmiEvictionBlockerDesc, prometheus.GaugeValue,
+		checkNonEvictableVMAndSetMetric(vmi),
+		vmi.Status.NodeName, vmi.Namespace, vmi.Name,
+	)
+	if err != nil {
+		return
+	}
+	ch <- mv
+
+}
+func updateVMIPhaseTransitionTime(vmi *k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+	if vmi.Status.PhaseTransitionTimestamp != nil {
+		transition := vmi.Status.PhaseTransitionTimestamp.Time
+		creation := vmi.CreationTimestamp.Time
+
+		diffSeconds := transition.Sub(creation).Seconds()
+
+		if diffSeconds < 0 {
+			// ignore if there is timestamp skew
+			return
+		}
+
 		mv, err := prometheus.NewConstMetric(
-			vmiEvictionBlockerDesc, prometheus.GaugeValue,
-			checkNonEvictableVMAndSetMetric(vmi),
-			nodeName, vmi.Namespace, vmi.Name,
+			vmiPhaseTransitionTimeDesc, prometheus.GaugeValue,
```

Comment:
> Is a ConstMetric a standalone metric? In other words, when prometheus scrapes this will it get a time for Pending and Scheduling or will it only see the most recent transition time and phase?

What's reported is the most recent transition time for a VMI at the point in time the metric is scraped. There are labels to this metric that allow selection by name, namespace, and phase.

So, Taking the average of `kubevirt_vmi_phase_transition_seconds_since_creation{phase=Running}` at any point in time would give you an indication of the average amount of time all currently running VMIs took to reach Running.  This could further be filtered by namespace and name. So if you started VMIs in different batches and wanted to separate the results, launching each batch in a separate namespace would be one way of doing that.

Reply:
The reason I'm asking is the time to running doesn't always tell the whole story.  For instance, if my PVCs were taking a long time to create for whatever reason, I would see a huge chunk of time spent in Pending.  But the `...{phase=Running}` metric doesn't always clue me into what the problem is, only that my time to running is a little longer than expected.  It sounds like though you can filter for `kubevirt_vmi_phase_transition_seconds_since_creation{phase=Pending}` and get that data, however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending.  I'm wondering if we can get a more granular look at the transition time.

Reply:
> The reason I'm asking is the time to running doesn't always tell the whole story. For instance, if my PVCs were taking a long time to create for whatever reason, I would see a huge chunk of time spent in Pending. But the ...{phase=Running} metric doesn't always clue me into what the problem is, only that my time to running is a little longer than expected. 

In prow I decide to record `last_phase` and `phase` https://github.com/kubernetes/test-infra/pull/14695 in the label to immediately indicate where it takes the time. It is then more verbose to get a sum of sub-states, but you can easily see where most of the time is spent and you can still get the overall-time instantly.

Either way I would suggest using a histogram and I would lean towards having `last_phast` and `phase` in the metrics.

> It sounds like though you can filter for kubevirt_vmi_phase_transition_seconds_since_creation{phase=Pending} and get that data, however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending. I'm wondering if we can get a more granular look at the transition time.

Two things on that:

  * In general I would be less interested in following individual VMIs. I would mostly want to see the histograms and compare the histogram with previous runs of the tests where e.g. 100VMs were started in parallel with well known VMI specs.
 * For specific VMI tracking: At mimimum for performance testing we should allow setting the VMI name/namespace on the metric. You will get the data for every transition if you choose the test-timing right. Basically something like this: Wait for a minute, add a unique test label to each metric on the prometheus config level, run the tests, wait for another minute. You will get all the data. When you then do queries (including sum, substractions, ...), you just have to choose the time-frame bigger than the test-window.
  * If we have long-running sub-tasks (like pvc provisioning) I think it would be worth making them explicitly visible in the metric.

Reply:
>Either way I would suggest using a histogram and I would lean towards having last_phast and phase in the metrics
>
Adding `last_phase` as a label sounds like it would help, however I still don't think that get us every detail.  I'm expecting a metric that has as its keys the VMs phase and Name, so the metric will have a record of each phase per VM.  I was thinking something like a [GaugeVec](https://pkg.go.dev/github.com/m3db/client_golang/prometheus#GaugeVec) (or [HistogramVec](https://pkg.go.dev/github.com/m3db/client_golang/prometheus#HistogramVec)).  Might look something like:

```go
	transition := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Namespace: ns,
			Subsystem: "phase",
			Name:      "transitionTime",
			Help:      "Phase transition time per VM.",
		},
		[]string{
			// Which VM we're measuring
			"vmName",
			// Which phase we measured
			"phase",
		},
	)
```

In Prometheus this should come out like:
```
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Pending”, vmName=“perf-1”}                 59s
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Scheduling”, vmName=“perf-1”}              60s
phase_transitionTime_gauge{namespace=“perf-test”, phase=“Running”, vmName=“perf-1”}                 75s
```

Reply:
I get the value of changing this metric to a histogram.  I'm unsure what tracking "last_phase" helps with though. 

Why wouldn't measuring each phase from creation time and putting that in a histogram expose the same data?

Reply:
>  however we would be dependent on how often prometheus scrapes e.g. 60s interval for a scrape and VMs spend 59s in Pending. I'm wondering if we can get a more granular look at the transition time.

@rthallisey I can avoid this by keeping a list of all phase transitions on the VMI status and reporting that in the histogram data.

When it comes to tracking the time, what do you think is most useful? Tracking each phase's duration anchored to creation time, or tracking each phase's duration anchored to the previous phases transition time?

Reply:
If we do time spent in each phase, then the server does the math and reports the result.  If we do time since create, then the client does the math.  I don't think either option closes off any use cases, so I like anchoring to the previous phase for convenience.

The one concern I have with anchoring to the previous phase is if the previous phase is missed.  For instance, can an edge triggered phase like `Scheduled` be missed?

>@rthallisey I can avoid this by keeping a list of all phase transitions on the VMI status and reporting that in the histogram data.
>
👍

Reply:
The advantage of having the interval between phases (`current_phase` - `last_phase`) is to apply aggregation (e.g. sum) in the metric.
If all phases have absolute time since creation, it will wrongly aggregate the time of all phases....

For example : 
`histogram_quantile(0.99,
   sum(irate(phase_transition_duration_miliseconds_bucket[1m]))
   by (vmName, le))
`

Otherwise, the query can become very complex to subtract the times....

Reply:
> The one concern I have with anchoring to the previous phase is if the previous phase is missed. For instance, can an edge triggered phase like Scheduled be missed?

yes it is possible. 

Ultimately what I want is two things

1. I want to know how long it takes until a VMI reaches running.
2. I want to know what phase we're spending the most time in.

I think the best way to solve this is two metrics. Keep what I have with creation to running, and add a histogram for transition times between phases as well.

Then we get everything. Someone can create a graph measuring just the time until Running and break down where we spend the most time in what phase.